process/block/preprocess/smartContractResults.go

```diff
@@ -0,0 +1,424 @@
+	"github.com/ElrondNetwork/elrond-go/hashing"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+type smartContractResults struct {
+	*basePreProcess
+	chRcvAllScrs                 chan bool
+	onRequestSmartContractResult func(shardID uint32, txHashes [][]byte)
+	scrForBlock                  txsForBlock
+	scrPool                      dataRetriever.ShardedDataCacherNotifier
+	storage                      dataRetriever.StorageService
+	scrProcessor                 process.SmartContractResultProcessor
+	accounts                     state.AccountsAdapter
+}
+
+// NewSmartContractResultPreprocessor creates a new smartContractResult preprocessor object
+func NewSmartContractResultPreprocessor(
+	scrDataPool dataRetriever.ShardedDataCacherNotifier,
+	store dataRetriever.StorageService,
+	hasher hashing.Hasher,
+	marshalizer marshal.Marshalizer,
+	scrProcessor process.SmartContractResultProcessor,
+	shardCoordinator sharding.Coordinator,
+	accounts state.AccountsAdapter,
+	onRequestSmartContractResult func(shardID uint32, txHashes [][]byte),
+) (*smartContractResults, error) {
+
+	if hasher == nil {
+		return nil, process.ErrNilHasher
+	}
+	if marshalizer == nil {
+		return nil, process.ErrNilMarshalizer
+	}
+	if scrDataPool == nil {
+		return nil, process.ErrNilScrDataPool
+	}
+	if store == nil {
+		return nil, process.ErrNilScrStorage
+	}
+	if scrProcessor == nil {
+		return nil, process.ErrNilTxProcessor
+	}
+	if shardCoordinator == nil {
+		return nil, process.ErrNilShardCoordinator
+	}
+	if accounts == nil {
+		return nil, process.ErrNilAccountsAdapter
+	}
+	if onRequestSmartContractResult == nil {
+		return nil, process.ErrNilRequestHandler
+	}
+
+	bpp := &basePreProcess{
+		hasher:           hasher,
+		marshalizer:      marshalizer,
+		shardCoordinator: shardCoordinator,
+	}
+
+	scr := &smartContractResults{
+		basePreProcess:               bpp,
+		storage:                      store,
+		scrPool:                      scrDataPool,
+		onRequestSmartContractResult: onRequestSmartContractResult,
+		scrProcessor:                 scrProcessor,
+		accounts:                     accounts,
+	}
+
+	scr.chRcvAllScrs = make(chan bool)
+	scr.scrPool.RegisterHandler(scr.receivedSmartContractResult)
+	scr.scrForBlock.txHashAndInfo = make(map[string]*txInfo)
+
+	return scr, nil
+}
+
+// waitForScrHashes waits for a call whether all the requested smartContractResults appeared
+func (scr *smartContractResults) waitForScrHashes(waitTime time.Duration) error {
+	select {
+	case <-scr.chRcvAllScrs:
+		return nil
+	case <-time.After(waitTime):
+		return process.ErrTimeIsOut
+	}
+}
+
+// IsDataPrepared returns non error if all the requested smartContractResults arrived and were saved into the pool
+func (scr *smartContractResults) IsDataPrepared(requestedScrs int, haveTime func() time.Duration) error {
```

Comment:
This method is not tested at all

Reply:
were copied from blockprocessor, not tested there with unit tests. Unit tests will be added. Added TODO

Reply:
ok